### PR TITLE
feat: foo20 check an error from grc20 package

### PR DESCRIPTION
**Pull request by r3v4s**
detailed infos are described in #857 

TL;DR
`foo20` relam doesn't check error from `grc20` package. 
- for example sending to address itself should fail, but currently runs fine

<details><summary>Checklists...</summary>

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](../.benchmarks/README.md).

## Maintainers Checklist

- [ ] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [ ] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [ ] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change
</details>

### CODE REVIEW

**Review by moul — APPROVED** [review]

**Comment by moul** [comment]
Related with #952

File: examples/gno.land/r/demo/foo20/foo20_test.gno [hunk 1]
```diff
@@ -54,3 +54,43 @@ func TestReadOnlyPublicMethods(t *testing.T) {
...
+		}
+		for _, tc := range tests {
+			shouldPanicWithMsg(t, tc.fn, tc.msg)
+		}
+	}
+}
+
+func a2u(addr std.Address) users.AddressOrName {
+	return users.AddressOrName(addr)
+}
```
Comment (harry-hov) [inline examples/gno.land/r/demo/foo20/foo20_test.gno:82]: I think this function is not needed. I prefer using `users.AddressOrName()` directly. 

If you prefer to keep it, maybe use a better name. And replace all occurrences of `users.AddressOrName()`

File: examples/gno.land/r/demo/foo20/foo20_test.gno [hunk 2]
```diff
@@ -54,3 +54,43 @@ func TestReadOnlyPublicMethods(t *testing.T) {
...
+func shouldPanicWithMsg(t *testing.T, f func(), msg string) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("The code did not panic")
+		} else {
+			errMsg := error(r).Error()
+			if errMsg != msg {
+				t.Errorf("excepted panic(%v), got(%v)", msg, errMsg)
+			}
+		}
```
Comment (harry-hov) [inline examples/gno.land/r/demo/foo20/foo20_test.gno:93]: Just a personal preference. You can ignore this comment if you want. 
```suggestion
		r := recover()
		if r == nil {
			t.Errorf("The code did not panic")
		}
		errMsg := error(r).Error()
		if errMsg != msg {
			t.Errorf("excepted panic(%v), got(%v)", msg, errMsg)
		}
```

**Review by harry-hov — APPROVED** [review]

File: examples/gno.land/r/demo/foo20/foo20_test.gno [hunk 3]
```diff
@@ -54,3 +54,43 @@ func TestReadOnlyPublicMethods(t *testing.T) {
 		}
 	}
 }
+
+func TestErrConditions(t *testing.T) {
+	admin := std.Address("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")
+	empty := std.Address("")
```
Comment (harry-hov) [inline examples/gno.land/r/demo/foo20/foo20_test.gno:60]: Maybe add another case?
```go
	invalid := std.Address("invalid")
```

Comment (r3v4s) [reply examples/gno.land/r/demo/foo20/foo20_test.gno:60]: grc20 package uses `checkIsValidAddress` to check whether address is valid or not.
https://github.com/gnolang/gno/blob/973be98226366b279ad2ab5931a53d08d0d6138f/examples/gno.land/p/demo/grc/grc20/util.gno#L7-L12

unfortunately it only checks `addr.String() == ""`, so that `invalid` won't be invalid but valid.

Comment (r3v4s) [reply examples/gno.land/r/demo/foo20/foo20_test.gno:60]: perhaps better to rename above function to `checkIsEmptyAddress`

File: examples/gno.land/r/demo/foo20/foo20_test.gno (hunk 1 above)
Comment (r3v4s) [reply examples/gno.land/r/demo/foo20/foo20_test.gno:82]: I'll follow origin style which removes `a2u()`

File: examples/gno.land/r/demo/foo20/foo20_test.gno (hunk 3 above)
Comment (waymobetta) [reply examples/gno.land/r/demo/foo20/foo20_test.gno:60]: Ethereum ecosystem libs like Ethers and web3.js commonly utilize the `isAddress` terminology, if the aim is for close parity or to minimize confusion errors for devs

Comment (moul) [reply examples/gno.land/r/demo/foo20/foo20_test.gno:60]: Let's handle this in the other refactor #952.

In this particular case, using `isAddress` doesn't seem suitable because we're dealing with an `std.Address` object and checking its validity. 

Creating a more general method like `std.Address{}.IsValid()` would be more appropriate in the future. For now, I'll focus on improving the check in the other PR or add a comment for clarification.

Thank you all for your suggestions.
